### PR TITLE
config validator, custom location and flag

### PR DESCRIPTION
**Pull request by tashima42**
* adds a config validator
* customizable config location
* updates default flags
* adds an option to input the config using a string

With these changes, automated tests will force us to maintain a valid example config and the generated example config won't replace the current one.

### CODE REVIEW

File: cmd/release/cmd/root.go [hunk 1]
```diff
@@ -21,9 +28,9 @@ var rootCmd = &cobra.Command{
 // Execute adds all child commands to the root command and sets flags appropriately.
 // This is called by main.main(). It only needs to happen once to the rootCmd.
 func Execute() {
+	cobra.OnInitialize(initConfig)
 	if err := rootCmd.Execute(); err != nil {
-		fmt.Println("error: " + err.Error())
-		os.Exit(1)
+		panic(err)
```
Comment (briandowns) [inline cmd/release/cmd/root.go:33]: Is the contents of the stack trace with a panic necessary here? Printing the error and exiting with a non 0 code is perferred.

File: cmd/release/cmd/root.go [hunk 2]
```diff
@@ -32,26 +39,40 @@ func SetVersion(version string) {
...
 func init() {
-	rootCmd.PersistentFlags().BoolP("debug", "d", false, "Debug")
-
-	configPath, err := config.DefaultConfigPath()
-	if err != nil {
-		fmt.Println(err)
-		os.Exit(1)
-	}
+	rootCmd.PersistentFlags().BoolVarP(&debug, "debug", "D", false, "Debug")
+	rootCmd.PersistentFlags().BoolVarP(&dryRun, "dry-run", "R", false, "Drun Run")
```
Comment (briandowns) [inline cmd/release/cmd/root.go:43]:
```suggestion
	rootCmd.PersistentFlags().BoolVarP(&dryRun, "dry-run", "R", false, "Dry Run")
```

File: cmd/release/cmd/root.go [hunk 3]
```diff
@@ -32,26 +39,40 @@ func SetVersion(version string) {
...
 
 func init() {
-	rootCmd.PersistentFlags().BoolP("debug", "d", false, "Debug")
-
-	configPath, err := config.DefaultConfigPath()
-	if err != nil {
-		fmt.Println(err)
-		os.Exit(1)
-	}
+	rootCmd.PersistentFlags().BoolVarP(&debug, "debug", "D", false, "Debug")
```
Comment (briandowns) [inline cmd/release/cmd/root.go:42]: What's the value / reason for changing the d to upper case?

File: cmd/release/config/config.go [hunk 4]
```diff
@@ -216,30 +162,30 @@ func exampleConfig() Config {
...
-				K3sPR:          "YOUR_TOKEN",
-				K3sPublish:     "YOUR_TOKEN",
-				RancherPR:      "YOUR_TOKEN",
-				RancherPublish: "YOUR_TOKEN",
-			},
 			GithubToken: "YOUR_TOKEN",
 			SSHKeyPath:  "path/to/your/ssh/key",
 		},
 	}
+	confBytes, err := json.MarshalIndent(conf, "", "  ")
```
Comment (briandowns) [inline cmd/release/config/config.go:184]:
```suggestion
	b, err := json.MarshalIndent(conf, "", "  ")
```

File: cmd/release/config/config.go [hunk 5]
```diff
@@ -102,89 +88,49 @@ type Config struct {
...
+func textEditorName() string {
+	editor := os.Getenv("EDITOR")
+	if editor == "" {
+		editor = "vi"
 	}
 
-	confB, err := json.MarshalIndent(exampleConfig(), "", " ")
+	return editor
+}
+func Load(configFile string) (*Config, error) {
```
Comment (briandowns) [inline cmd/release/config/config.go:109]: Is this function or the one below used outside of this package?

File: cmd/release/cmd/root.go (hunk 3 above)
Comment (tashima42) [reply cmd/release/cmd/root.go:42]: There isn't any big value on changing it, the main reason is to maintain consistency with the dry run flag. Neither are being used by any sub-cmds yet

File: cmd/release/config/config.go (hunk 5 above)
Comment (tashima42) [reply cmd/release/config/config.go]: Yes, they are both used by the `cmd` package

File: cmd/release/cmd/root.go [hunk 6]
```diff
@@ -21,9 +28,9 @@ var rootCmd = &cobra.Command{
 // Execute adds all child commands to the root command and sets flags appropriately.
 // This is called by main.main(). It only needs to happen once to the rootCmd.
 func Execute() {
+	cobra.OnInitialize(initConfig)
 	if err := rootCmd.Execute(); err != nil {
-		fmt.Println("error: " + err.Error())
-		os.Exit(1)
+		log.Fatal(err)
```
Comment (briandowns) [inline cmd/release/cmd/root.go:33]: Will this add additional info to the output? If so I'd prefer we keep the output only what we need.

File: cmd/release/cmd/root.go [hunk 7]
```diff
@@ -32,26 +39,40 @@ func SetVersion(version string) {
...
-	if err != nil {
-		fmt.Println("failed to load config, use 'release config gen' to create a new one at: " + configPath)
-		fmt.Println(err)
-		os.Exit(1)
+	var conf *config.Config
+	var err error
+	if stringConfig != "" {
+		conf, err = config.Read(strings.NewReader(stringConfig))
+		if err != nil {
+			panic(err)
```
Comment (briandowns) [inline cmd/release/cmd/root.go:60]: Is the stack trace necessary here since we know why we're erroring?

**Review by briandowns — DISMISSED** [review]

File: cmd/release/cmd/root.go (hunk 6 above)
Comment (tashima42) [reply cmd/release/cmd/root.go]: Updated, `rootCmd.Execute()` already prints the error, printing it again duplicates the output

**Review by nicholasSUSE — CHANGES_REQUESTED** [review]

**Review by nicholasSUSE — APPROVED** [review]

**Review by brooksn — APPROVED** [review]

**Pull request merged**